File: backend/internal/model/household.go
package model

import (
	"time"

	"gorm.io/gorm"
)

// Household roles. Only owners can invite, change grace period, or dissolve.
const (
	RoleOwner       = "owner"
	RoleContributor = "contributor"
	RoleViewOnly    = "view_only"
)

// Account share visibility levels. Absence of an account_shares row = "private".
// VisibilityPrivate is an API-level value: sending it on PUT clears the share.
// It is never stored — the CHECK constraint only permits the other two.
const (
	VisibilityPrivate        = "private"
	VisibilityBalanceOnly    = "balance_only"
	VisibilityBalanceAndTxns = "balance_and_txns"
)

type Household struct {
	ID   int64  `gorm:"primaryKey" json:"id"`
	Name string `gorm:"not null" json:"name"`
	// Ownership is not stored here. The single source of truth is the
	// household_members row with role='owner' (one active owner per
	// household, enforced by uq_household_single_owner). See #283.
	GracePeriodDays int            `gorm:"not null;default:30" json:"grace_period_days"`
	CreatedAt       time.Time      `json:"created_at"`
	UpdatedAt       time.Time      `json:"updated_at"`
	DeletedAt       gorm.DeletedAt `gorm:"index" json:"-"`
}

func (Household) TableName() string { return "households" }

// HouseholdMember tracks a user's membership lifecycle. left_at marks the start
// of grace; purged_at marks final removal. See ADR-0007.
type HouseholdMember struct {
	ID          int64      `gorm:"primaryKey" json:"id"`
	HouseholdID int64      `gorm:"not null" json:"household_id"`
	UserID      int64      `gorm:"not null" json:"user_id"`
	Role        string     `gorm:"not null" json:"role"`
	JoinedAt    time.Time  `json:"joined_at"`
	LeftAt      *time.Time `json:"left_at,omitempty"`
	PurgedAt    *time.Time `json:"purged_at,omitempty"`
}

func (HouseholdMember) TableName() string { return "household_members" }

type HouseholdInvite struct {
	ID          int64      `gorm:"primaryKey" json:"id"`
	HouseholdID int64      `gorm:"not null" json:"household_id"`
	TokenHash   string     `gorm:"not null" json:"-"`
	Role        string     `gorm:"not null" json:"role"`
	CreatedBy   int64      `gorm:"not null" json:"created_by"`
	ExpiresAt   time.Time  `gorm:"not null" json:"expires_at"`
	AcceptedAt  *time.Time `json:"accepted_at,omitempty"`
	AcceptedBy  *int64     `json:"accepted_by,omitempty"`
	CreatedAt   time.Time  `json:"created_at"`
}

func (HouseholdInvite) TableName() string { return "household_invites" }

type AccountShare struct {
	ID          int64          `gorm:"primaryKey" json:"id"`
	AccountID   int64          `gorm:"not null" json:"account_id"`
	HouseholdID int64          `gorm:"not null" json:"household_id"`
	Visibility  string         `gorm:"not null" json:"visibility"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
	DeletedAt   gorm.DeletedAt `gorm:"index" json:"-"`
}

func (AccountShare) TableName() string { return "account_shares" }

// Household-owned budgets and goals are no longer separate models: ADR-0018
// folded shared_budgets/shared_goals into the unified Budget / SavingsGoal
// tables, owned via HouseholdID. See repository.PlanOwner.

File: frontend/src/types/household.ts
// Mirror of backend service/household types. Keep aligned with
// backend/internal/model/household.go and service/household/service.go.

export type HouseholdRole = 'owner' | 'contributor' | 'view_only'

export type Household = {
  id: number
  name: string
  // Ownership is not a household field — it's the member with role 'owner'
  // (single source of truth, backend #283).
  grace_period_days: number
  created_at: string
  updated_at: string
}

export type HouseholdMember = {
  id: number
  household_id: number
  user_id: number
  role: HouseholdRole
  joined_at: string
  left_at?: string | null
  purged_at?: string | null
}

// HouseholdDetail mirrors service/household.HouseholdDetail.
export type HouseholdDetail = {
  household: Household
  members: HouseholdMember[]
  role: HouseholdRole // requester's role in this household
}

// MembersListing mirrors service/household.MembersListing — the listing
// endpoint returns active members + (optionally) in-grace members.
export type MembersListing = {
  active: HouseholdMember[]
  in_grace?: HouseholdMember[] | null
}

export type HouseholdInvite = {
  id: number
  household_id: number
  role: HouseholdRole
  created_by: number
  expires_at: string
  accepted_at?: string | null
  accepted_by?: number | null
  created_at: string
}

// CreateInviteResult mirrors service/household.CreateInviteResult — the
// raw token is only returned at creation time.
export type CreateInviteResult = {
  invite: HouseholdInvite
  token: string
}

// SharedBudget mirrors backend model.SharedBudget.
export type SharedBudget = {
  id: number
  household_id: number
  category_id: number
  period: 'monthly' | 'weekly' | 'annual'
  amount: string
  rollover: boolean
  is_active: boolean
  created_at: string
  updated_at: string
}

export type CreateSharedBudgetInput = {
  category_id: number
  period: 'monthly' | 'weekly' | 'annual'
  amount: string
  rollover?: boolean
  is_active?: boolean
}

export type UpdateSharedBudgetInput = {
  category_id?: number
  period?: 'monthly' | 'weekly' | 'annual'
  amount?: string
  rollover?: boolean
  is_active?: boolean
}

// SharedGoal mirrors backend model.SharedGoal.
export type SharedGoal = {
  id: number
  household_id: number
  name: string
  target_amount: string
  current_amount: string
  target_date?: string | null
  created_at: string
  updated_at: string
}

export type CreateSharedGoalInput = {
  name: string
  target_amount: string
  target_date?: string // YYYY-MM-DD
}

export type UpdateSharedGoalInput = {
  name?: string
  target_amount?: string
  target_date?: string
  clear_target_date?: boolean
}
